### ISSUE TITLE
Clip_gradient setting in gluon.trainer doesn’t work

### ISSUE THREAD

**Issue by zhaolili** [apache/incubator-mxnet#13512]
[Setting clip gradients in Trainer doesn’t work.](https://discuss.mxnet.io/t/clip-gradient-setting-in-gluon-trainer-doesnt-work/2471)
**mxnet version: 1.2.0**

Below code is used to clip gradients to [-2.0, 2.0]. With this setting, I printed the weights gradients of the last convolutional layer, and it turns out that some gradients are still much larger than 2.0.

```python 
trainer = gluon.Trainer(ssh_net.collect_params(), ‘sgd’, {‘learning_rate’:1e-4, ‘wd’:5e-4, ‘clip_gradient’: 2., ‘lr_scheduler’:lr_sch, ‘momentum’:0.9})
```

**Comment by evi-Genius** [comment]
hi, I also meet this problem. Do you have sovled this problem?

**Comment by onurcanbektas** [comment]
Same here!

**Comment by szha** [comment]
the gradient clipping happens inside optimizer update and is not updated back to the gradient buffers.

**Comment by onurcanbektas** [comment]
@szha Thank you very much for the answer; how do you use the `clip_global_norm` with a hybridized network?

**Comment by onurcanbektas** [comment]
@szha Also, `clip_global_norm` function does a rescaling and not a clipping. We just want want to ignore the gradients which are larger than some threshold.

**Comment by szha** [comment]
> @szha Also, `clip_global_norm` function does a rescaling and not a clipping. We just want want to ignore the gradients which are larger than some threshold.

I see. So you want to clip the absolute value of each dimension in the gradients instead of rescale based on global norm. You can do the following in this case:
```python
for p in net.collect_params().values():
    if p.grad_req != 'null':
        for g in p.list_grad():
            mx.nd.clip(g, MIN, MAX, out=g)
```

**Comment by onurcanbektas** [comment]
Thanks a lot for the help @szha ; I've tried your proposed solution and done many different simulations but in each case, the problem persisted (though the gradient explosion is deferred to later stages of the learning)
